### ISSUE TITLE
Backup "Delete old backups" fragility

### ISSUE THREAD

**Issue by swelljoe** [virtualmin/virtualmin-gpl#151]
It's not obvious how to make "Delete old backups" work, and it fails pretty much silently when is isn't configured correctly. e.g. https://forum.virtualmin.com/t/solved-backup-virtual-servers-old-backups-not-deleted-from-s3/60634

I've done it, too...I don't think I could get it right without several attempts.

**Comment by chris001** [comment]
Code fixes needed?

**Comment by jcameron** [comment]
Looks like the specific fix needed in that case is to handle a suffix like _daily

**Comment by ftrojahn** [comment]
I see another problem here. Virtualmin backup seems to rely on the file system date of the directories.
Cause of that, the backups of one server fail from time to time due to full backup space.

While the files within the backup dir have the correct date ...

```
:/var/local/vmbackups# ls -l 04-02/
...
-rw------- 1 root root            1208 Apr  2 04:43 shop.domain.tld.tar.gz
-rw------- 1 root root  331232898 Apr  2 04:43 shop.domain.tld.tar.gz
....
-rw------- 1 root root          15660 Apr  2 05:41 virtualmin.tar.gz
-rw-r--r-- 1 root root              154 Apr  2 05:41 virtualmin.tar.gz.info
```
... the directories have not:

```
:/var/local/vmbackups# ls -lrt
insgesamt 68
drwx------ 2 root root 4096 Apr 14 12:41 04-02
drwx------ 2 root root 4096 Apr 14 12:41 04-04
drwx------ 2 root root 4096 Apr 14 12:41 04-05
drwx------ 2 root root 4096 Apr 14 12:41 04-06
drwx------ 2 root root 4096 Apr 14 12:41 04-07
drwx------ 2 root root 4096 Apr 14 12:41 04-08
drwx------ 2 root root 4096 Apr 14 12:41 04-09
drwx------ 2 root root 4096 Apr 15 05:13 04-15
drwx------ 2 root root 4096 Apr 16 05:06 04-16
drwx------ 2 root root 4096 Apr 17 04:51 04-17
drwx------ 2 root root 4096 Apr 17 18:03 04-14
drwx------ 2 root root 4096 Apr 17 18:03 04-13
drwx------ 2 root root 4096 Apr 17 18:03 04-12
drwx------ 2 root root 4096 Apr 17 18:03 04-10
drwx------ 2 root root 4096 Apr 18 05:35 04-18
drwx------ 2 root root 4096 Apr 19 05:34 04-19
drwx------ 2 root root 4096 Apr 20 05:17 04-20

:/var/local/vmbackups-belenus# ls -lc
insgesamt 68
drwx------ 2 root root 4096 Apr 14 12:41 04-02
drwx------ 2 root root 4096 Apr 14 12:41 04-04
drwx------ 2 root root 4096 Apr 14 12:41 04-05
drwx------ 2 root root 4096 Apr 14 12:41 04-06
drwx------ 2 root root 4096 Apr 14 12:41 04-07
drwx------ 2 root root 4096 Apr 14 12:41 04-08
drwx------ 2 root root 4096 Apr 14 12:41 04-09
drwx------ 2 root root 4096 Apr 17 18:03 04-10
drwx------ 2 root root 4096 Apr 17 18:03 04-12
drwx------ 2 root root 4096 Apr 17 18:03 04-13
drwx------ 2 root root 4096 Apr 17 18:03 04-14
drwx------ 2 root root 4096 Apr 15 05:13 04-15
drwx------ 2 root root 4096 Apr 16 05:06 04-16
drwx------ 2 root root 4096 Apr 17 04:51 04-17
drwx------ 2 root root 4096 Apr 18 05:35 04-18
drwx------ 2 root root 4096 Apr 19 05:34 04-19
drwx------ 2 root root 4096 Apr 20 05:17 04-20

```
So, virtualmin finds nothing to delete:
```
Deleting backups from local file /var/local/vmbackups/%m-%d older than 14 days ..
.. found 17 backups, but none were old enough
```
Here, /var/local is an nfs mounted directory made available by the provider. I don't know,
why these directorie's times differ. Apart from removing some of the backups to get
some space, we did not change anything manually.

**Comment by PitWenkin** [comment]
`Here, /var/local is an nfs mounted directory made available by the provider. I don't know,
why these directorie's times differ. Apart from removing some of the backups to get
some space, we did not change anything manually.`

Once you manually remove files inside a backup folder the timestamp of the whole folder is updated to the "last change"-date.
I had the same experience on a NFS mounted disk

**Comment by ftrojahn** [comment]
yes, you‘re right - I forgot, that one of the vhosts’s backups were deleted at certain times from specific dirs.

But anyway: Virtualmin backup knows, from the format of the strftime vars, which backup was created at what time. So should it really rely on directory dates? Just my opinion.

**Comment by PitWenkin** [comment]
The date in the name of the backup folder is the date of the beginning of the backup, while the date of the directory itself is the date of the last change. If nothing messed with the date this should be date date where the backup was completed.

Depending on the volume of the backup an the time it began, this could be to different dates.
If it takes multiple days to complete it could be that some backups that are still running are deleted …

**Comment by ftrojahn** [comment]
> Depending on the volume of the backup an the time it began, this could be to different dates.
> If it takes multiple days to complete it could be that some backups that are still running are deleted …
ok, I agree - that makes sense.

**Comment by jcameron** [comment]
The current code doesn't try to compute the date from the filename - it assumes that the filesystem timestamp is correct. This is because sometimes the filename doesn't contain enough info to figure out the date, for example if it was just named after the day of the week.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
